### PR TITLE
Private/Pre ICO dates & rate changes

### PR DESCRIPTION
**Pull request by jamesmorgan**
* [x] ability to specify private/pre ICO dates - only modifiable by owner
  * [x] tests
* [x] Ability to provide private/pre rates with
  * [x] tests
* [x] ability to change rates when thresholds pass
  * [x] tests
* [x] Update Pauseable tests
* [x] ready for review

### CODE REVIEW

File: contracts/PixieCrowdsale.sol [hunk 1]
```diff
@@ -32,6 +32,62 @@ contract PixieCrowdsale is CappedCrowdsale, WhitelistedCrowdsale, IndividualLimi
...
+    // TODO should there be any validation on rate?
+
+    privateSaleRate = _rate;
+    privateSaleCloseTime = _privateSaleCloseTime;
+  }
+
+  /**
+   * @dev sets the pre sale close time and rate
+   */
+  function setPreSaleCloseTime(uint256 _preSaleCloseTime, uint256 _rate) public onlyOwner {
```
Comment (andygray) [inline contracts/PixieCrowdsale.sol]: this implies these need to called in a certain order. Can we set them both at once which would ensure it is all or nothing?

File: contracts/PixieCrowdsale.sol [hunk 2]
```diff
@@ -32,6 +32,62 @@ contract PixieCrowdsale is CappedCrowdsale, WhitelistedCrowdsale, IndividualLimi
   RefundableCrowdsale(_goal) {
   }
 
+  uint256 public privateSaleCloseTime;
```
Comment (andygray) [inline contracts/PixieCrowdsale.sol]: In the end contract we could hard code these in theory like some other ICOs do - but no good for now

File: contracts/PixieCrowdsale.sol [hunk 3]
```diff
@@ -32,6 +32,62 @@ contract PixieCrowdsale is CappedCrowdsale, WhitelistedCrowdsale, IndividualLimi
...
+    preSaleRate = _rate;
+    preSaleCloseTime = _preSaleCloseTime;
+  }
+
+  /**
+   * @dev Overridden method used to allow different rates for private/pre sale
+   * @param _weiAmount Value in wei to be converted into tokens
+   * @return Number of tokens that can be purchased with the specified _weiAmount
+   */
+  function _getTokenAmount(uint256 _weiAmount) internal view returns (uint256) {
```
Comment (andygray) [inline contracts/PixieCrowdsale.sol]: Is it worth sticking all this in a superclass called `PrivatePresalePricingStructure`

**Comment by andygray** [comment]
`Expose ability to modify private/pre dates in web app` << Would be nice to set all this in the constructor - shame about max vars

File: contracts/PixieCrowdsale.sol (hunk 3 above)
Comment (jamesmorgan) [reply contracts/PixieCrowdsale.sol]: Ye thats on my list of refactorings

File: contracts/PixieCrowdsale.sol (hunk 2 above)
Comment (jamesmorgan) [reply contracts/PixieCrowdsale.sol]: The good thing about this is that it means that in theory you can alter the end dates but hard coding is an option

File: contracts/PixieCrowdsale.sol (hunk 1 above)
Comment (jamesmorgan) [reply contracts/PixieCrowdsale.sol]: They could be together - left speerate for easy for now i.e. not sure if needed  or if only one is needed etc.

File: contracts/PixieCrowdsale.sol (hunk 3 above)
Comment (jamesmorgan) [reply contracts/PixieCrowdsale.sol]: Also may be easier to leave here as it uses rate and is Ownable so the mico contract will have to inherit from a few others, we can tweak if needs be.

File: contracts/PixieCrowdsale.sol [hunk 4]
```diff
@@ -55,6 +105,14 @@ contract PixieCrowdsale is CappedCrowdsale, WhitelistedCrowdsale, IndividualLimi
   */
   function _preValidatePurchase(address _beneficiary, uint256 _weiAmount) internal {
     super._preValidatePurchase(_beneficiary, _weiAmount);
+
+    // enforce private/pre sale rates to be set
```
Comment (andygray) [inline contracts/PixieCrowdsale.sol]: and times

**Pull request merged**